File: yt/yt/flow/examples/java/word_count/wordcount/src/main/java/tech/ytsaurus/flow/examples/wordcount/WordCountMapper.java
package tech.ytsaurus.flow.examples.wordcount;

import org.springframework.stereotype.Component;
import tech.ytsaurus.flow.computation.OutputCollector;
import tech.ytsaurus.flow.context.RuntimeContext;
import tech.ytsaurus.flow.examples.wordcount.model.Word;
import tech.ytsaurus.flow.examples.wordcount.model.WordCountState;
import tech.ytsaurus.flow.function.RowFunction;
import tech.ytsaurus.flow.row.ExtendedMessage;
import tech.ytsaurus.flow.state.InternalStateDescriptor;
import tech.ytsaurus.flow.state.StateAccessor;
import tech.ytsaurus.flow.state.StateDescriptors;

@Component
public class WordCountMapper implements RowFunction {
    static final InternalStateDescriptor<WordCountState> WORD_STATE =
            StateDescriptors.yson("word-state", WordCountState.class);

    // [BEGIN on_message]
    @Override
    public void onMessage(ExtendedMessage message, OutputCollector output, RuntimeContext ctx) {
        Word input = message.getPayload();
        StateAccessor<WordCountState> stateAccessor = ctx.getState(WORD_STATE, message);
        var state = stateAccessor.getOrDefault(new WordCountState(input.getWord(), 0));
        state.setCount(state.getCount() + 1);
        stateAccessor.set(state);
    }
    // [END on_message]
}

File: yt/yt/flow/examples/kotlin/word_count/wordcount/src/main/kotlin/tech/ytsaurus/flow/examples/wordcount/WordCountMapper.kt
package tech.ytsaurus.flow.examples.wordcount

import org.springframework.stereotype.Component
import tech.ytsaurus.flow.computation.OutputCollector
import tech.ytsaurus.flow.context.RuntimeContext
import tech.ytsaurus.flow.examples.wordcount.model.Word
import tech.ytsaurus.flow.examples.wordcount.model.WordCountState
import tech.ytsaurus.flow.function.RowFunction
import tech.ytsaurus.flow.row.ExtendedMessage
import tech.ytsaurus.flow.state.InternalStateDescriptor
import tech.ytsaurus.flow.state.StateDescriptors

@Component
class WordCountMapper : RowFunction {
    companion object {
        val WORD_STATE: InternalStateDescriptor<WordCountState> =
            StateDescriptors.yson("word-state", WordCountState::class.java)
    }

    // [BEGIN on_message]
    override fun onMessage(message: ExtendedMessage, output: OutputCollector, ctx: RuntimeContext) {
        val input: Word = message.getPayload()
        val stateAccessor = ctx.getState(WORD_STATE, message)
        val state = stateAccessor.getOrDefault(WordCountState(input.word, 0))
        state.count += 1
        stateAccessor.set(state)
    }
    // [END on_message]
}
